[WIP] Add sklearn's HistGradientBoosting

ahmedkrmn opened the pull request:
Closes #64:

- This PR adds sklearn's [HistGradientBoostingClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.HistGradientBoostingClassifier.html) to hummingbird.

- It modifies the two functions `convert_sklearn_gbdt_classifier()` and `get_parameters_for_sklearn_common()` to handle the HistGradientBoostingClassifier model attributes.

- It also updates the documentation to include the new HistGradientBoostingClassifier.

ahmedkrmn committed:
Add sklearn's HistGradientBoosting

ahmedkrmn committed:
Add sklearn's HistGradientBoosting

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/signed)](https://cla.opensource.microsoft.com/microsoft/hummingbird?pullRequest=87) <br/>All CLA requirements met.

ahmedkrmn commented:
@interesaaat getting the value of Alpha is what's left.

@NicolasHug I checked your [comment](https://github.com/microsoft/hummingbird/issues/11#issuecomment-636325188) which references using the `_baseline_prediction` attribute of the HistGradientBoostingClassifier model for doing this.
I noticed that it switches between being a `<class 'numpy.ndarray'>` and `<class 'numpy.float64'>` during different iterations, so could you elaborate on that?

NicolasHug commented:
Right my bad, it's not always a float. In multiclass problems it's a numpy array of size `n_classes` because we build `n_classes` trees at each iteration. The decision function for a given class corresponds to the sum of all the predictions of the trees of that class, and the predicted probability to belong to a class is a softmax of these sums. It is the same idea as in multinomial logistic regression, where we get `n_classes` weight vectors.

Since we build `n_classes` set of trees, we need to compute one initial prediction per class.

ahmedkrmn commented:
@NicolasHug OK great, thanks a lot for this info!

@interesaaat I'm still not sure about how we are going to map the `_baseline_prediction` to alpha and also how `init_` is being mapped to alpha is still vague for me. For example, why do we have to take the natural logarithm here?
```python
alpha = [[np.log(operator.raw_operator.init_.class_prior_[i]) for i in range(n_classes)]]
```

interesaaat commented on hummingbird/ml/operator_converters/_tree_commons.py:
```diff
@@ -169,20 +169,42 @@ def get_tree_params_and_type(tree_infos, get_tree_parameters, extra_config):
+        values = [trees.nodes[x]["value"] for x in range(len(trees.nodes))]
+        # Convert to numpy.ndarray with shape (s, 1, 1)
+        values = np.array(values)
+        values = values.reshape((values.shape[0], 1, 1))
```
Not sure why you need to this. Isn't reshape`(-1, 1)` enough?

ahmedkrmn replied:
`values` has a shape of `(x, 1, 1)` in case of HistGradientBoosting.
Doing `values.reshape(-1, 1)` we will have a shape of `(x, 1)` instead of `(x, 1, 1)`. If this won't cause any problems I can modify it.

interesaaat replied:
Yes, can you please put them as `-1, 1`?

interesaaat commented on hummingbird/ml/supported.py:
```diff
@@ -34,11 +35,15 @@ def _build_sklearn_operator_list():
     """
     if sklearn_installed():
+        # Enable experimental to import HistGradientBoostingClassifier
+        from sklearn.experimental import enable_hist_gradient_boosting
```
Why is this needed?

NicolasHug replied:
it's because the HistGBDT estimators are still experimental and are not subject to the 2-versions deprecation cycles, so we requires users to explicitly say "I want to use this feature anyway"

When the estimators become stable, this will have no effect

interesaaat reviewed (commented):
Looks good. I haven't test it yet tho. I left a couple of comments. We will need to fix the coverage, but let's finish the coding first.

interesaaat commented:
> @NicolasHug OK great, thanks a lot for this info!
> 
> @interesaaat I'm still not sure about how we are going to map the `_baseline_prediction` to alpha and also how `init_` is being mapped to alpha is still vague for me. For example, why do we have to take the natural logarithm here?
> 
> ```python
> alpha = [[np.log(operator.raw_operator.init_.class_prior_[i]) for i in range(n_classes)]]
> ```

I believe that something like `alpha = np.array([operator.raw_operator._baseline_prediction.flatten().tolist()])` should work. The `init_` part is not clear to me either. @NicolasHug is the `init_` here used as in GradientBoosting? In that case we can just do a check as treelite (as pointed by Nicolas [here](https://github.com/microsoft/hummingbird/issues/11#issuecomment-636325188)) and open as issue if the value is different than `zero`. In this way we can have the operator in and figure out later this corner case.

NicolasHug commented:
**Short version**

> I believe that something like alpha = np.array([operator.raw_operator._baseline_prediction.flatten().tolist()]) should work

Yes that should work (modulo shape conversion). As far as I understand, reading the existing code in `convert_sklearn_gbdt_classifier`, `ALPHA` corresponds to the initial prediction of the boosting process, which is precisely `_baseline_prediciton`.


**Long version**

> is the init_ here used as in GradientBoosting?

it is used in the GradientBoostingXXX estimators but not in the HistGradientBoostingXXX estimator. 

In the GradientBoostingXXX estimators, we allow users to pass an `init` parameter which is an arbitrary estimator object that specifies the very first prediction, i.e. the value at which predictions are initialized when starting the boosting process. (Kinda like the value you would use to initialize weights when doing gradient descent)

In the HistGBDT code, there is no `init` parameter and we just hardcode it. The specific value isn't super important, but it corresponds to the mean of the target in regression (with LS loss), and to the log-odds ratio in binary classification, because that's the "best"  we can do w.r.t. the loss while only looking at the targets.

Predicting the log-odds ratio is also the default when `init=None` in the GradientBoostingXXX estimators, though in this case a `dummy` estimator is used. That dummy estimator always predicts a constant prediction which is the prior of the positive class, stored as its `class_prior_[1]` attribute.

The current code for the GradientBoostingClassifier is correct: when `init=None`, `ALPHA` is set to the log-odds ratio to replicate the behavior of the estimator 
(nice job to whoever wrote this BTW, this isn't obvious!)

```py
    # Get the value for Alpha.
    if operator.raw_operator.init == "zero":
        alpha = [[0.0]]
    elif operator.raw_operator.init is None:
        if n_classes == 1:
            alpha = [[np.log(operator.raw_operator.init_.class_prior_[1] / (1 - operator.raw_operator.init_.class_prior_[1]))]]
        else:
            alpha = [[np.log(operator.raw_operator.init_.class_prior_[i]) for i in range(n_classes)]]
    else:
        raise RuntimeError("Custom initializers for GBDT are not yet supported in Hummingbird."
```

This is for the GradientBoostingClassifier. For the HistGradientBoosting estimators, there is no dummy-estimator, and the log-odds ratio is just in `_baseline_prediction`. Basically `_baseline_prediction` corresponds to `np.log(operator.raw_operator.init_.class_prior_[1] / (1 - operator.raw_operator.init_.class_prior_[1])`, so you just have to replace that

interesaaat commented:
> The current code for the GradientBoostingClassifier is correct: when `init=None`, `ALPHA` is set to the log-odds ratio to replicate the behavior of the estimator
> (nice job to whoever wrote this BTW, this isn't obvious!)

@scnakandala is the genius behind everything :) 

Thanks for the detailed explanation @NicolasHug! Perhaps renaming `Alpha` to `base_prediction` or `base_score` (as in XGBoost) will be a bit more clear. WDYT?

ahmedkrmn committed:
Implement alpha for HistGradientBoosting

ahmedkrmn committed:
Implement alpha for HistGradientBoosting

interesaaat reviewed (changes requested):
Looks good! 2 asks:
- Can you please add in this PR the tests Karla wrote for you?
- Can you please merge the changes we pushed into master?

Will merge after these are fixed (plus the reshape).

ahmedkrmn committed:
Merge branch 'master' of https://github.com/microsoft/hummingbird into hist-GBDT

ahmedkrmn commented:
Hi @interesaaat,
I've added the test by @ksaur. Unfortunately, it fails on 9 out of the 11 tests.
I'll push the changes so that you can have a look on the failing output.

ahmedkrmn committed:
Add HistGBDT test by Karla Saur

ahmedkrmn committed:
Add HistGBDT test by Karla Saur

interesaaat commented:
> Hi @interesaaat,
> I've added the test by @ksaur. Unfortunately, it fails on 9 out of the 11 tests.
> I'll push the changes so that you can have a look on the failing output.

I see. Apparently the probabilities are off for some reason. Were the tests passing on your side before commit  0f82328?

ahmedkrmn commented:
> > Hi @interesaaat,
> > I've added the test by @ksaur. Unfortunately, it fails on 9 out of the 11 tests.
> > I'll push the changes so that you can have a look on the failing output.
> 
> I see. Apparently the probabilities are off for some reason. Were the tests passing on your side before commit [0f82328](https://github.com/microsoft/hummingbird/commit/0f8232879b90bce1d4aa77b6df89d7b741d92d39)?

Yeah I tried before merging, they failed as well.

KranthiGV commented on tests/test_sklearn_histgbdt_converters.py:
```diff
@@ -0,0 +1,85 @@
+        X = np.random.rand(10, 1)
+        X = np.array(X, dtype=np.float32)
+        # y = np.random.randint(2, size=10)
+
```
nit: Did we miss the actual test here? Currently, it doesn't test any. Can we please add any assertions or remove this method?

KranthiGV commented:
> > > Hi @interesaaat,
> > > I've added the test by @ksaur. Unfortunately, it fails on 9 out of the 11 tests.
> > > I'll push the changes so that you can have a look on the failing output.
> > 
> > 
> > I see. Apparently the probabilities are off for some reason. Were the tests passing on your side before commit [0f82328](https://github.com/microsoft/hummingbird/commit/0f8232879b90bce1d4aa77b6df89d7b741d92d39)?
> 
> Yeah I tried before merging, they failed as well.

Looking at the test results, the predicted probability is off by a high value (many a times even the first decimal digit is different while the expected decimal precision in test is 6). That's why all the 9 tests that assert on predicted probabilities fail.

Did they ever pass locally for you @ahmedkrmn? We can backtrack using that info. Also, is the predicted class correct?

Also, interesaaat / ksaur, how do you people generally debug these issues?

interesaaat commented:
> Also, interesaaat / ksaur, how do you people generally debug these issues?

Well, we start with regression, 1 tree and see why the results don't match. Then we do regression multiple trees -> binary -> multiclass. In the worst case we check the partial results within Sklearn execution and where they start to differ wrt Hummingbird.

I can take a look at this.

ahmedkrmn commented:
Please let me know if there is anything I can do to help.

On Thu, Jun 4, 2020, 6:14 PM Matteo Interlandi <notifications@github.com>
wrote:

> Also, interesaaat / ksaur, how do you people generally debug these issues?
>
> Well, we start with regression, 1 tree and see why the results don't
> match. Then we do regression multiple trees -> binary -> multiclass. In the
> worst case we check the partial results within Sklearn execution and where
> they start to differ wrt Hummingbird.
>
> I can take a look at this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/microsoft/hummingbird/pull/87#issuecomment-638957139>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE7MCZTJERD5BQPPGW3HX53RU7B4XANCNFSM4NO2PGFA>
> .
>

ahmedkrmn commented:
@interesaaat any new updates?

interesaaat commented:
No sorry I have been working in the ONNX PR. I am done with it so I will look into this next.

interesaaat committed:
fix bug with the conversion

interesaaat committed:
fix bug with the conversion

interesaaat commented:
Hey @ahmedkrmn  I fixed the problem with the conversion. Basically, learning rate and normalization for `hist_gbdt` are already computed upfront and we don't need to do them on our side. 

2 questions:
- can you please merge with master (sorry)
- can you strip the `hist_gbdt` implementation from the `gbdt` one? At the beginning they were quite similar, but now they differ quite a bit so I prefer to have separate implementations. You can even just add the code in `_tree_commons.py` into `gbdt.py` at this point.

ahmedkrmn commented:
> Hey @ahmedkrmn I fixed the problem with the conversion. Basically, learning rate and normalization for `hist_gbdt` are already computed upfront and we don't need to do them on our side.
> 
> 2 questions:
> 
> * can you please merge with master (sorry)
> * can you strip the `hist_gbdt` implementation from the `gbdt` one? At the beginning they were quite similar, but now they differ quite a bit so I prefer to have separate implementations. You can even just add the code in `_tree_commons.py` into `gbdt.py` at this point.

Great Matteo, thanks for the update!
I'll use a separate `convert_sklearn_hist_gbdt_classifier` function for dealing with histGBDT.

Regarding moving the code of `_tree_commons.py` to `gbdt.py`, I can see that `get_parameters_for_sklearn_common()` function is used in`decision_tree.py`, so is moving it a good idea?
Or what you mean is that we create a separate `get_parameters_for_sklearn_common()` for histGBDT and put it in `gbdt.py` and leave the original function as is in the `_tree_commons.py`?

interesaaat commented:
> > Hey @ahmedkrmn I fixed the problem with the conversion. Basically, learning rate and normalization for `hist_gbdt` are already computed upfront and we don't need to do them on our side.
> > 2 questions:
> > 
> > * can you please merge with master (sorry)
> > * can you strip the `hist_gbdt` implementation from the `gbdt` one? At the beginning they were quite similar, but now they differ quite a bit so I prefer to have separate implementations. You can even just add the code in `_tree_commons.py` into `gbdt.py` at this point.
> 
> Great Matteo, thanks for the update!
> I'll use a separate `convert_sklearn_hist_gbdt_classifier` function for dealing with histGBDT.
> 
> Regarding moving the code of `_tree_commons.py` to `gbdt.py`, I can see that `get_parameters_for_sklearn_common()` function is used in`decision_tree.py`, so is moving it a good idea?
> Or what you mean is that we create a separate `get_parameters_for_sklearn_common()` for histGBDT and put it in `gbdt.py` and leave the original function as is in the `_tree_commons.py`?

Sorry, I only meant the code related to `hist_gbdt`m like from like 183 in `gbdt`. The other side of the condition is ok if stays in `_tree_commons.py`

ahmedkrmn commented:
Hi @interesaaat,
Does `test_sklearn_histgbdt_converters.py` pass on your local machine after your additions?
It's still failing on mine.

interesaaat commented:
> Hi @interesaaat,
> Does `test_sklearn_histgbdt_converters.py` pass on your local machine after your additions?
> It's still failing on mine.

Yes it is passing all tests in my machine. Are you getting the same failures or something different? Like, all tests are still failing?

ahmedkrmn commented on hummingbird/ml/operator_converters/_tree_commons.py:
```diff
@@ -180,22 +180,11 @@ def get_parameters_for_sklearn_common(tree_infos):
+        features = [n[2] if n[2]!=0 else 0 for n in trees.nodes]
+        thresholds = [n[3] if n[3]!=0 else -1 for n in trees.nodes]
+        lefts = [n[4] if n[4]!=0 else -1 for n in trees.nodes]
+        rights = [n[5] if n[5]!=0 else -1 for n in trees.nodes]
```
Hi @interesaaat , the test is failing because you've used wrong indices for accessing `lefts` and `rights`.
The index of `lefts` is `5` and for `rights` it's `6`.
I suggest accessing by the string name of the property instead of the index to avoid such bugs.

interesaaat replied:
Agreed! I think that we probably have different sklearn versions, and they changed the indexing between versions.

ahmedkrmn committed:
Create a separate converter function for histGBDT

- Create a separate tree parser for histGBDT.
- Revert the changes made to get_parameters_for_sklearn_common().

ahmedkrmn commented:
@interesaaat When running the tests locally, sometimes one test fails as in the case with this last failing build. If you rerun the tests again, all of them pass, so that's a strange behavior!

interesaaat commented:
Can you please point me to the failing case? Is it the multiclass? This is probably due to rounding errors and we can fix it momentarily by increasing the tolerance.

ahmedkrmn commented:
> Can you please point me to the failing case? Is it the multiclass? This is probably due to rounding errors and we can fix it momentarily by increasing the tolerance.

I ran the tests 20 times. It's always either that they all pass or 9/10 pass.

**Summary:**
10/20 all passed.
02/20 failed at `test_HistGBDT_classifier_binary_converter`.
02/20 failed at `test_HistGBDT_perf_tree_trav_multi_classifier_converter`.
02/20 failed `test_HistGBDT_gemm_classifier_converter`.
01/20 failed at `test_HistGBDT_tree_trav_multi_classifier_converter`.
01/20 failed at `test_HistGBDT_shifted_labels_converter`.

interesaaat commented:
I can take a look. Can you please rebase with master? After my pass we can then merge this PR!

ahmedkrmn commented:
> I can take a look. Can you please rebase with master? After my pass we can then merge this PR!

Done!

interesaaat committed:
Some refactoring and decrease of tolerance

interesaaat committed:
fix style errors

interesaaat committed:
use a fixed test dataset

ahmedkrmn commented:
Thanks a lot, @interesaaat,
This was really an interesting PR to work on!
